include/event_sources/timer.h

```diff
@@ -28,12 +28,57 @@ enum class TimerType
   INTERVAL,
 };
 
+class BaseTimer : public virtual microloop::EventSource
+{
+public:
+  /**
+   * How many times the timer has expired so far.
+   */
+  virtual std::uint64_t get_expirations_count() const = 0;
+
+  virtual ~BaseTimer()
+  {}
+};
+
+class TimerController
+{
+public:
+  TimerController(BaseTimer *timer, microloop::EventLoop *event_loop) :
+      timer{timer}, event_loop{event_loop}
+  {}
+
+  /**
+   * Cancel the timer.
+   */
+  void cancel()
+  {
+    event_loop->remove_event_source(timer);
+  }
+
+  /**
+   * How many times the timer has expired so far.
+   * @return [description]
+   */
+  std::uint64_t get_expirations_count() const
+  {
+    return timer->get_expirations_count();
+  }
+
+private:
+  BaseTimer *timer;
+  microloop::EventLoop *event_loop;
+};
+
 template <class Callback>
```

Comment:
Maybe add a `static_assert` that checks `std::is_invocable` for `Callback`?
https://en.cppreference.com/w/cpp/types/is_invocable

Reply:
Good point.